File: python/Problem014.py
#
# This file contains a solution to the problem number 14 of the Project Euler, last modified on 5/13/2018.
# Copyright (c) Matthieu Le Boucher. All rights reserved.
#

def result():
    MAX_SIZE = 1_000_000
    max_chain_length = 1
    max_chain_number = 1

    chain_lengths = [0] * MAX_SIZE
    chain_lengths[1] = 1

    for i in range(2, MAX_SIZE):
        n = i
        current_length = 1

        while n != 1:
            if n % 2 == 0:
                n //= 2

                if n < i:
                    chain_lengths[i] = current_length + chain_lengths[n]

                    if chain_lengths[i] > max_chain_length:
                        max_chain_length = chain_lengths[i]
                        max_chain_number = i

                    break
            else:
                n = 3 * n + 1

            current_length += 1

    return max_chain_number

if __name__ == "__main__":
    print(result())

File: cpp/Problem014.cpp
#include "EulerSolution.h"
#include <iostream>

/*
 * @author Matthieu Le Boucher <matt.leboucher@gmail.com>
 */

using namespace std;

#define MAX_STARTING_NUMBER 1000000

class Problem014 : public EulerSolution {
public:
    int number() override {
        return 14;
    }

    double solution() override {
        int maxChainLength = 1;
        int maxChainNumber = 1;

        int* chainLengths = new int[MAX_STARTING_NUMBER];
        chainLengths[1] = 1;

        int currentChainLength;
        long i;

        for (int n = 500000; n < MAX_STARTING_NUMBER; n++) {
            // Compute current chain length.
            i = n;
            currentChainLength = 1;

            while (i != 1) {
                if (i % 2 == 0) {
                    i /= 2;

                    // If Collatz(i) has already been seen.
                    if (i < n) {
                        chainLengths[n] = currentChainLength + chainLengths[i];

                        if (chainLengths[n] > maxChainLength) {
                            maxChainLength = chainLengths[n];
                            maxChainNumber = n;
                        }

                        break;
                    }
                } else
                    i = 3 * i + 1;

                currentChainLength++;
            }
        }

        return maxChainNumber;
    }

    double answer() override {
        return 837799;
    }
};